Implement PKCE-based OAuth2 with InPost API

miho120 opened the pull request:
## 📦 Direct InPost OAuth2 Authentication and API Requests

### Summary

This release introduces a major architectural change by replacing the external authentication service with direct InPost OAuth2 authentication. The integration now communicates directly with InPost's official API, improving reliability and security.

Closes #12

---

### ✨ What's New

#### Authentication Overhaul
- **Direct OAuth2 Flow:** Replaced external authentication service with InPost's official OAuth2 implementation
- **PKCE Support:** Implemented Proof Key for Code Exchange (PKCE) for enhanced security
- **Token Management:** Access and refresh tokens are now stored and managed locally
- **Email Verification:** Added support for email confirmation step (required for existing InPost accounts)
- **Multi-language Support:** Authentication flow now supports Polish and English (default)

#### New Sensors
- **Locker Address:** New sensor displaying full locker location (city, street, building number)
- **Locker Description:** New sensor with location hints (e.g., "near Biedronka store")

---

### 📚 Documentation

- Updated README with practical usage examples:
  - Dashboard panel setup with Markdown card
  - Parcel pickup notification automation

---

### 🛠️ Code Quality Improvements

- Increased test coverage
- Organized exceptions into dedicated `exceptions` module
- Extracted HTTP client logic into separate `http_client` module
- Refactored utility functions into `utils` module

---

### ⚠️ Breaking Changes

> **Important:** Existing users will need to re-configure the integration after updating.

| Change | Impact |
|--------|--------|
| OAuth2 authentication | External authentication service is no longer used |
| Config entry structure | Data structure changed; existing configurations are incompatible |

#### Migration Steps

> **Note:** Simply re-adding the integration should work for most users. Full removal (steps 1-2) is only needed if you encounter issues.

1. Remove the existing InPost Paczkomaty integration from Home Assistant
2. Restart Home Assistant
3. Add the integration again and complete the new authentication flow
4. Re-select your preferred parcel lockers

---

### 🧪 Testing

- [x] Fresh installation tested
- [x] Authentication flow (SMS + email verification) tested
- [x] Token refresh mechanism tested
- [x] All sensors populate correctly
- [x] Existing automations work with new entity structure

---

### 📋 Checklist

- [x] Code follows project style guidelines
- [x] Tests added/updated
- [x] Documentation updated
- [x] CHANGELOG updated
- [x] Breaking changes documented

miho120 commented:
Hi @jakon89 👋
Could you please review this PR and release a new version after it's merged?

jakon89 commented on README.md:
```diff
@@ -16,7 +16,7 @@ This Home Assistant integration tracks your parcels by fetching data from InPost
    which you also provide.
 2. **Data Flow:** This authentication data is passed to a backend server (hosted by @jakon89), which then uses it to
```
Could you please update the step descriptions of steps 1 and 2?

jakon89 commented on custom_components/inpost_paczkomaty/api.py:
```diff
@@ -1,154 +1,376 @@
+            url=f"{API_BASE_URL}{self.PROFILE_ENDPOINT}",
+            custom_headers={
+                # without this InPost API returns 500 Internal Server Error
+                "User-Agent": "InPost-Mobile/4.4.2 (1)-release (iOS 26.2; iPhone15,3; pl)",
```
Could you please move it to CONST? This is repeated a second time in this file and we will need to bump it often I guess.

jakon89 commented on custom_components/inpost_paczkomaty/coordinator.py:
```diff
@@ -7,29 +7,43 @@
             hass,
             _LOGGER,
             name="InPost Paczkomaty data coordinator",
             update_interval=timedelta(seconds=30),
```
Under the hood backend polls InPost API every 300s because of rate limiting - can we meet somewhere closer to 300 than 30?

miho120 replied:
In my opinion no need. I executed a lot of requests and haven't actually been rate limited. Just now I've sent 100 requests within 10 seconds and wasn't rate limited. This integration makes a lot less requests. Therefore, in my opinion we can leave it as it is.

In future we can migrate this value to a config file, so users can set it by theirselves.

jakon89 commented on custom_components/inpost_paczkomaty/translations/pl.json:
```diff
@@ -1,32 +1,46 @@
+      "invalid_phone_format": "Nieprawidłowy format numeru telefonu. Podaj dokładnie 9 cyfr.",
+      "phone_unknown_server_error": "Wystąpił nieznany błąd serwera podczas rejestracji numeru telefonu.",
+      "invalid_code": "Podany kod jest nieprawidłowy lub wygasł.",
+      "rate_limited_error": "Zbyt wiele żądań. Spróbuj ponownie później.",
```
Is this still used? I don't see any usages 🤔

jakon89 commented on README.md:
```diff
@@ -43,7 +44,56 @@ This Home Assistant integration tracks your parcels by fetching data from InPost
 4. Execute steps **6, 7, and 8** from the HACS installation method above.
 
----
+## Usage Examples
```
Do you think it would be beneficial to add a banner on the top of README to indicate that `0.3.0` has major changes and the integration re-install might be needed?

miho120 replied:
Added 
<img width="1007" height="371" alt="image" src="https://github.com/user-attachments/assets/215b76ba-3687-4d99-ad9c-fdd5d0ab97d0" />

jakon89 reviewed (changes requested):
I dind't expect such a PR! Hats off for the work! I added some minor comments - please check.

The backend uses a completely different flow/endpoints (much simpler IMO), but I am ok with this flow if it works. Is it (at least partially) an official API that I am not aware of?

Once the comments are adressed I will merge the PR and release a new version. Please give me a day or two to test it on my test HA instance.

Thanks a lot!

jakon89 commented on pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [project]
 name = "inpost-paczkomaty"
-version = "0.2.0"
+version = "0.3.0"
```
This version is auto-updated by the CI pipeline on release - please revert.

miho120 commented on CHANGELOG.md:
```diff
@@ -1,3 +1,36 @@
+# [0.3.0](https://github.com/jakon89/Inpost-Paczkomaty/compare/v0.2.0...v0.3.0) (2025-12-29)
```
Commit messages aren't my straight part 😅
Anyway, reverting the file

miho120 commented on custom_components/inpost_paczkomaty/inpost_auth_flow.py:
```diff
@@ -0,0 +1,360 @@
+
+        Raises:
+            IdentityAdditionLimitReachedError: When identity addition limit reached.
+            PhoneNumberAlreadyRegisteredError: When phone is already registered.
```
Good catch. Leftovers, removed.

miho120 commented:
> I dind't expect such a PR! Hats off for the work! I added some minor comments - please check.
> 
> The backend uses a completely different flow/endpoints (much simpler IMO), but I am ok with this flow if it works. Is it (at least partially) an official API that I am not aware of?
> 
> Once the comments are adressed I will merge the PR and release a new version. Please give me a day or two to test it on my test HA instance.
> 
> Thanks a lot!

No worries, I really love this integration when it used to work, so I got notifications about parcels on the way home and not when I'm at home already from my wife 😅 Therefore after it stopped working for me I decided just not to wait. It's a community effort anyway 😀

I haven't found any official documentation for non-business customers. This flow is a result of revert engineering the InPost mobile app and debugging HTTP traffic it makes to understand the flow and the API they use in mobile app.

Thank you @jakon89  for you comments! I addressed them.

Cheers!
